File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int i,j,n=s1.size(),m;
        vector<int>cnt(26,0),f(26,0);
        for(i=0;i<n;i++)
        {
            cnt[s1[i]-'a']++;
        }
        m=s2.size();
        for(i=0;i<m;i++)
        {
            f[s2[i]-'a']++;
            if(i+1>=n)
            {
                bool flag=true;
                for(j=0;j<26;j++)
                {
                    if(cnt[j]==f[j])
                        continue;
                    else{ flag=false;break;}
                }
                if(flag) return true;
                f[s2[i-n+1]-'a']--;
            }
        }
        return false;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        hashMap = {chr(ch): 0 for ch in range(ord('a'), ord('z') + 1)}
        for ch in s1:
            hashMap[ch]+=1
        last = -1
        temp = hashMap.copy()
        for i,ch in enumerate(s2):
            if temp[ch]>=1:
                temp[ch]-=1
                if last ==-1:
                    last=i
            else:
                if ch not in s1:
                    last=-1
                    temp=hashMap.copy()
                elif last!=-1:
                    while temp[ch]<=0:
                        temp[s2[last]]+=1
                        last+=1
                    temp[ch]-=1
                    
            if last!=-1 and (i-last+1)==len(s1):
                return True
        return False
        